NPE when parsing a lambda-invoking function with `bodyOnly = true`

Given some function like:

```kotlin
fun myFunction() = runBlocking {
  println("boo!")
}
```

and a rule like:

```kotlin
rule("my-rule") {
  replacement = sourceCode(
    fqName = "com.example.MyClass.myFunction",
    bodyOnly = true
  )
}
```

We'd get an exception like:

```
Caused by: java.lang.NullPointerException
        at com.rickbusarow.doks.internal.psi.NamedSamples.findAll(NamedSamples.kt:143)
        at com.rickbusarow.doks.internal.psi.NamedSamples.findAll(NamedSamples.kt:106)
        at com.rickbusarow.doks.DoksParseTask.execute(DoksParseTask.kt:77)
        [...]
```

which is this line:
```kotlin
namedDeclaration.bodyBlockExpression!!.textInScope()
```

The logic is looking for a body in the function itself, whereas that function is using expression syntax.  The thing that looks to be the "body" is actually the body of the lambda, whereas the body of the function is null.

--- a/doks-gradle-plugin/src/main/kotlin/com/rickbusarow/doks/internal/psi/NamedSamples.kt
+++ b/doks-gradle-plugin/src/main/kotlin/com/rickbusarow/doks/internal/psi/NamedSamples.kt
@@ -44,6 +44,7 @@ import org.jetbrains.kotlin.psi.psiUtil.getChildrenOfType
 import org.jetbrains.kotlin.psi.psiUtil.parents
 import java.io.File
 import kotlin.LazyThreadSafetyMode.NONE
+import kotlin.contracts.contract
 import kotlin.io.path.Path
 
 @Serializable
@@ -118,21 +119,23 @@ internal class NamedSamples(
       val content =
         when (val namedDeclaration = cache[request.fqName]) {
           is KtClassOrObject -> if (request.bodyOnly) {
-            namedDeclaration.body!!.textInScope()
+            namedDeclaration.body
+              .requireBodyNotNull(namedDeclaration, request.fqName) {
+                "A type declaration must have a body"
+              }
+              .textInScope()
           } else {
             namedDeclaration.text.trimIndentAfterFirstLine()
           }
 
           is KtProperty -> if (request.bodyOnly) {
             namedDeclaration.initializer
-              .requireNotNull {
-                "${namedDeclaration.containingKtFile} > " +
-                  "A property must have an initializer when using 'bodyOnly = true'."
+              .requireBodyNotNull(namedDeclaration, request.fqName) {
+                "A property must have an initializer"
               }
               .let { (it as? KtStringTemplateExpression) ?: it.getChildOfType() }
-              .requireNotNull {
-                "${namedDeclaration.containingKtFile} > " +
-                  "A property initializer must be a string template."
+              .requireBodyNotNull(namedDeclaration, request.fqName) {
+                "A property initializer must be a string template"
               }
               .textInScope()
           } else {
@@ -140,7 +143,11 @@ internal class NamedSamples(
           }
 
           is KtNamedFunction -> if (request.bodyOnly) {
-            namedDeclaration.bodyBlockExpression!!.textInScope()
+            namedDeclaration.bodyBlockExpression
+              .requireBodyNotNull(namedDeclaration, request.fqName) {
+                "A function must use body syntax"
+              }
+              .textInScope()
           } else {
             namedDeclaration.text.trimIndentAfterFirstLine()
           }
@@ -153,6 +160,26 @@ internal class NamedSamples(
         }
 
       SampleResult(request, content)
+    }
+  }
+
+  inline fun <T : Any> T?.requireBodyNotNull(
+    psi: KtElement,
+    fqName: String,
+    message: () -> String
+  ): T {
+    contract {
+      returns() implies (this@requireBodyNotNull != null)
+    }
+    return requireNotNull(this) {
+      """
+      |${psi.containingKtFile.absolutePath()} > ${message()} when using 'bodyOnly = true'.
+      |requested name: $fqName
+      |matched code:
+      |---
+      |${psi.text.trimIndentAfterFirstLine()}
+      |---
+      """.trimMargin()
     }
   }
 

--- a/doks-gradle-plugin/src/main/kotlin/com/rickbusarow/doks/internal/stdlib/string.kt
+++ b/doks-gradle-plugin/src/main/kotlin/com/rickbusarow/doks/internal/stdlib/string.kt
@@ -133,7 +133,13 @@ internal fun String.remove(vararg regex: Regex): String = regex.fold(this) { acc
 
 internal fun String.trimIndentAfterFirstLine(): String {
   val split = lines()
+
+  // If there's only one line, there's nothing to trim,
+  // and the remaining logic would just append an extra newline.
+  if (split.size == 1) return this
+
   val first = split.first()
-  val remaining = split.drop(1).joinToString("\n").trimIndent()
+  val remaining = split.drop(1).joinToString("\n")
+    .trimIndent()
   return "$first\n$remaining"
 }

--- a/doks-gradle-plugin/src/test/kotlin/com/rickbusarow/doks/internal/psi/NamedSamplesTest.kt
+++ b/doks-gradle-plugin/src/test/kotlin/com/rickbusarow/doks/internal/psi/NamedSamplesTest.kt
@@ -16,6 +16,7 @@
 package com.rickbusarow.doks.internal.psi
 
 import io.kotest.assertions.throwables.shouldNotThrowAny
+import io.kotest.assertions.throwables.shouldThrow
 import io.kotest.matchers.shouldBe
 import org.intellij.lang.annotations.Language
 import org.junit.jupiter.api.Test
@@ -240,6 +241,95 @@ class NamedSamplesTest {
   }
 
   @Test
+  fun `an class without a body and bodyOnly = true`() {
+
+    // test for: https://github.com/rickbusarow/Doks/issues/223
+
+    val exception = shouldThrow<IllegalArgumentException> {
+      parse(
+        fqName = "com.test.MyClass",
+        bodyOnly = true,
+        """
+      package com.test
+
+      class MyClass
+      """
+      )
+    }
+
+    exception.message shouldBe """
+      src/Source_0.kt > A type declaration must have a body when using 'bodyOnly = true'.
+      requested name: com.test.MyClass
+      matched code:
+      ---
+      class MyClass
+      ---
+    """.trimIndent()
+  }
+
+  @Test
+  fun `an expression syntax function that invokes a lambda and bodyOnly = true`() {
+
+    // test for: https://github.com/rickbusarow/Doks/issues/223
+
+    val exception = shouldThrow<IllegalArgumentException> {
+      parse(
+        fqName = "com.test.MyClass.myFunction",
+        bodyOnly = true,
+        """
+      package com.test
+
+      class MyClass {
+        fun myFunction() = runBlocking {
+          println("boo!")
+        }
+      }
+      """
+      )
+    }
+
+    exception.message shouldBe """
+      src/Source_0.kt > A function must use body syntax when using 'bodyOnly = true'.
+      requested name: com.test.MyClass.myFunction
+      matched code:
+      ---
+      fun myFunction() = runBlocking {
+        println("boo!")
+      }
+      ---
+    """.trimIndent()
+  }
+
+  @Test
+  fun `an expression syntax function that does not invoke lambda and bodyOnly = true`() {
+
+    // test for: https://github.com/rickbusarow/Doks/issues/223
+
+    val exception = shouldThrow<IllegalArgumentException> {
+      parse(
+        fqName = "com.test.MyClass.myFunction",
+        bodyOnly = true,
+        """
+      package com.test
+
+      class MyClass {
+        fun myFunction() = Unit
+      }
+      """
+      )
+    }
+
+    exception.message shouldBe """
+      src/Source_0.kt > A function must use body syntax when using 'bodyOnly = true'.
+      requested name: com.test.MyClass.myFunction
+      matched code:
+      ---
+      fun myFunction() = Unit
+      ---
+    """.trimIndent()
+  }
+
+  @Test
   fun `a variable inside a nested function can be resolved`() {
 
     parse(
@@ -370,7 +460,6 @@ class NamedSamplesTest {
       """
     ) shouldBe """
       |object Foo
-      |
     """.trimMargin()
   }
 
@@ -379,7 +468,7 @@ class NamedSamplesTest {
       DoksPsiFileFactory()
         .createKotlin(
           name = "Source_$index.kt",
-          path = "Source_$index.kt",
+          path = "src/Source_$index.kt",
           content = code.trimIndent()
         )
     }

